bluetooth: host: Add APIs to track Unenhanced ATT MTU

ubieda opened the pull request:
### Description
This PR features an additional API to get the current UATT MTU and a callback on UATT MTU changes. This in order to provide the ability to keep track of this parameter, regardless of the presence of additional EATT channels.

Fixes #68070.

### Changes
- Add `uatt_mtu_updated` to get notified on UATT MTU changes on a given connection.
- Add `bt_gatt_get_uatt_mtu` to get a connection's current UATT MTU.
- Update MTU Update sample to use the above.
- Refactor Babblesim test: MTU Update to have peripheral and central be two separate builds.

### Testing
- Compiled all BabbleSim Testsuite. Ran Bluetooth Host Testsuite locally.
- Built MTU Update Peripheral Sample.

ubieda commented:
@alwa-nordic @jori-nordic @Thalley @carlescufi @jhedberg I'd like your inputs on this approach before going any further.

This solution seems the simplest one without:
- Breaking existing code that is relying on `att_mtu_updated`.
- Depending on the BT Stack refactoring involving inclusion of L2CAP channels in the API parameters (#68107).

I know there were mixed feelings on naming it "UATT". That is also up for discussion, provided the approach is OK.

Before taking this PR out of draft state, I'm planning to add Babblesim tests and/or samples supporting this feature (similar to #67858).

alwa-nordic commented:
A new callback is fine by me.

alwa-nordic commented:
The spec defines the ATT MTU of a bearer as a single value that's the same for rx and tx. See `bt_att_mtu`. I think it may be best to give this value in the callback to minimize user errors.

We may need to specify more precisely what the values in the callback mean, so that the behavior does not change with implementation details. Maybe like it's the value of the "Client Rx MTU" and the "Server Rx MTU" fields in the MTU exchange PDU, where the locally controlled value is `rx`?

The only use I can think of for separate rx and tx MTU is to allow the application to detect that the local side is the bottleneck, which is useful, but then maybe we should instead give the application the final ATT MTU and the remote mtu. Can you think of other use cases?

carlescufi commented:
Bluetooth WG:

- Remove rx and tx and instead use att_mtu
- Add the missing `bt_uatt_get_mtu` getter

ubieda committed:
bluetooth: host: Add API to get a connection's UATT MTU

Similar to bt_att_get_mtu but only considering the Unenhanced ATT
channel.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: bsim: bluetooth: Refactor MTU Update test

To isolate Central and Peripheral in two separate builds.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: bluetooth: mtu_update: Modified Sample to use UATT MTU

To demonstrate the usage of the UATT MTU API to determine if a payload
with its associated length can be sent over a GATT notification.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
doc: releases: bluetooth: host: Add entry of UATT MTU APIs

Referencing bt_gatt_get_uatt_mtu added, marked as experimental.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
> Bluetooth WG:
> 
> * Remove rx and tx and instead use att_mtu
> * Add the missing `bt_uatt_get_mtu` getter
> * This is a short-term solution. The long-term one is to replace `bt_conn` with `bt_l2cap_chan` in all ATT/GATT calls.

I've addressed the agreed changes. PR is up for review now.

alwa-nordic commented:
Should the new API be labeled experimental?

Also, we should take a moment to discover and discuss in detail the exact behavior of this new API, so that we are sure we know what we are making.

I have some observations which I think are worth discussing:

 - `uatt_mtu_updated` will always fire at connection establishment with the value `23`.
 - `uatt_mtu_updated` will always fire after a exchange even if the value does not change.
 - There can be two exchanges, and `uatt_mtu_updated` will fire for each, even though the value will be the same.
 - It's not possible to deduce whether the MTU exhange has happened using `bt_att_get_uatt_mtu`.
 - The exchange can happen at almost any moment, so the value of `bt_att_get_uatt_mtu` can be stale.

I'll withhold my comments about alternatives for a moment, so we can all consider it for ourselves.

When we are happy with the behavior, etc, we should document it all, and ideally test it too.

jori-nordic commented:
dropping callback ok w/ me

jori-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1371,6 +1380,17 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 
+/** @brief Get UATT MTU for a connection
+ *
+ *  Get negotiated UATT connection MTU, note that this does not equal the largest
```
This could return the mtu pre-negotiation. I think you can drop the `negotiated`. 

The second part of the sentence is also a bit unclear to me: what exactly are you referring to by `packet` and `attribute data`. Is it an on-air data packet, or L2CAP PDU?

(we were reviewing at my desk w/ alwa-nordic, maybe he has a suggestion)

ubieda replied:
Sorry for the confusion. I based this comment on the existing `bt_gatt_get_mtu`:
https://github.com/zephyrproject-rtos/zephyr/blob/878640fa0817a5c6684758d68dcb824e24571612/include/zephyr/bluetooth/gatt.h#L1480-L1489

I've updated this API description following both yours and @alwa-nordic's suggestion.

jori-nordic commented on tests/bsim/bluetooth/host/att/mtu_update/test_scripts/run_test.sh:
```diff
@@ -5,13 +5,14 @@
 set -eu
 source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
 
-simulation_id="$(guess_test_long_name)"
```
can't comment on a deleted file. Could you revert `_compile.sh`? You can rename it to `compile.sh` to align with @alwa-nordic 's https://github.com/zephyrproject-rtos/zephyr/pull/76956

jori-nordic reviewed (changes requested):
the only NAKs are 
- compile.sh
- getter's docstring

the rest is just nitpicks, you can ignore if you don't want to address.

alwa-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1371,6 +1380,17 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 
+/** @brief Get UATT MTU for a connection
+ *
+ *  Get negotiated UATT connection MTU, note that this does not equal the largest
+ *  amount of attribute data that can be transferred within a single packet.
```
> note that this does not equal the largest amount of attribute data that can be transferred within a single packet.

This is unclear. How about:

> The ATT_MTU defines the largest size of an ATT PDU, encompassing the ATT opcode, additional fields, and any attribute value payload. Consequently, the maximum size of a value payload is less and varies based on the type of ATT PDU. For example, in an ATT_HANDLE_VALUE_NTF PDU, the Attribute Value field can contain up to ATT_MTU - 3 octets.

alwa-nordic commented:
More context to think about:

A function to get the current ATT_MTU is not useful to determine where to truncate a notification because the ATT_MTU could change before the notification is sent, and whether the ATT PDU maxes out the ATT_MTU is relevant to the protocol:

Core 3.F.3.4.7.1 "ATT_HANDLE_VALUE_NTF":
> If the attribute value is longer than (ATT_MTU-3) octets, then only the first (ATT_MTU-3) octets of this attributes value can be sent in a notification.
>
> Note: For a client to get a long attribute, it must use the ATT_READ_BLOB_REQ PDU (see Section 3.4.4.5) following the receipt of this notification

ubieda commented:
I've updated the PR based on the latest feedback:
- Dropped Callback API.
- Brought back `compile.sh` script and tested it's working.
- Modified Getter API docstring to reflect suggestions.

LMK if you have further comments!

jori-nordic commented on samples/bluetooth/mtu_update/peripheral/src/peripheral_mtu_update.c:
```diff
@@ -69,6 +100,14 @@ void run_peripheral_sample(uint8_t *notify_data, size_t notify_data_size, uint16
+		} else {
+			printk("Skipping notification since UATT MTU is not sufficient."
+			       "Required: %d, Actual: %d\n",
+			       notify_data_size + 3,
```
you still have some `+ 3`s

ubieda replied:
Apologies, should be covered now.

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -3832,6 +3832,27 @@ uint16_t bt_att_get_mtu(struct bt_conn *conn)
+
+	/** There must be an UATT channel per connection */
+	k_oops();
+	CODE_UNREACHABLE;
```
We should avoid `CODE_UNREACHABLE`. It's very unsafe (a better name would be `INVOKE_UNDEFINED_BEHAVIOR`), so we need a good reason for using it.

```suggestion
	return 0;
```

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -3832,6 +3832,27 @@ uint16_t bt_att_get_mtu(struct bt_conn *conn)
+	}
+
+	/** There must be an UATT channel per connection */
+	k_oops();
```
It the comment above is an assertion, then it's correct to put an `__ASSERT_NO_MSG(false)` here.

OTOH, if the comment above is something you are unsure about (I don't blame you -- I don't think it's obvious e.g. during connection teardown), then it's better to try recovering. Returning `0`, if we document that, or the default ATT_MTU value may be better.

@jori-nordic What do you think?

alwa-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1488,6 +1488,22 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
+ *
+ *  @param conn Connection object.
+ *
+ *  @return MTU in bytes
```
It looks like the implementation can return `0` in some cases. We should document this.

alwa-nordic replied:
For example:
```
@retval 0 If @p conn does not have an UATT ATT_MTU, for example because it's disconnected.
@retval >0 The UATT ATT_MTU at this moment.
```

ubieda replied:
Sounds good. Documented this retval explicitly

Thalley commented on tests/bsim/bluetooth/host/att/mtu_update/src/main_peripheral.c:
```diff
@@ -0,0 +1,73 @@
+
+static void test_peripheral_main(void)
+{
+	notify_data[13] = 0x7f;
+	notify_data[99] = 0x55;
```
Why these specific values? Are they checked anywhere?

Thalley replied:
:magic_wand:

ubieda commented on samples/bluetooth/mtu_update/peripheral/src/peripheral_mtu_update.c:
```diff
@@ -11,11 +11,15 @@
 #define MTU_TEST_SERVICE_TYPE BT_UUID_128_ENCODE(0x2e2b8dc3, 0x06e0, 0x4f93, 0x9bb2, 0x734091c356f0)
 
+/* Overhead: opcode (u8) + handle (u16) */
+#define ATT_PACKET_SIZE(payload_len) (1 + 2 + payload_len)
```
Agreed - Thanks for the suggestion!

ubieda commented on tests/bsim/bluetooth/host/att/mtu_update/src/common.h:
Sounds good. However, since there are many instances with these duplicate defines, may we handle it on a separate PR?

Thalley replied:
Since you are just moving code, it's fine as is :)

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1488,6 +1488,23 @@ bool bt_gatt_is_subscribed(struct bt_conn *conn,
 
+/** @brief Get Unenhanced ATT (UATT) MTU for a connection
+ *
+ *  Get UATT connection MTU.
+ *
+ * The ATT_MTU defines the largest size of an ATT PDU, encompassing the ATT
+ * opcode, additional fields, and any attribute value payload. Consequently,
+ * the maximum size of a value payload is less and varies based on the type
+ * of ATT PDU. For example, in an ATT_HANDLE_VALUE_NTF PDU, the Attribute Value
+ * field can contain up to ATT_MTU - 3 octets (size of opcode and handle).
+ *
+ *  @param conn Connection object.
```
```suggestion
 *  Get UATT connection MTU.
 *
 *  The ATT_MTU defines the largest size of an ATT PDU, encompassing the ATT
 *  opcode, additional fields, and any attribute value payload. Consequently,
 *  the maximum size of a value payload is less and varies based on the type
 *  of ATT PDU. For example, in an ATT_HANDLE_VALUE_NTF PDU, the Attribute Value
 *  field can contain up to ATT_MTU - 3 octets (size of opcode and handle).
 *
 *  @param conn Connection object.
```

Minor; inconsistency in spacing

ubieda replied:
I see. Let me take care of that.

ubieda replied:
Done!

ubieda commented:
There are some conflicts. I'll need to rebase
